node/src/types/block.rs

```diff
@@ -884,6 +898,90 @@ impl BlockBody {
             .unwrap_or_else(|error| panic!("should serialize block body: {}", error));
         hash::hash(&serialized_body)
     }
+
+    /// Computes the body hash by hashing a Merkle tree of the form:
+    ///
+    /// ```text
+    ///                   body_hash
+    ///                   /      \__________________
+    /// hash(deploy_hashes)      /                  \_____________
+    ///                   hash(transfer_hashes)     /             \
+    ///                                         hash(proposer)   SENTINEL
+    /// ```
+    fn hash_v2(&self) -> Digest {
+        let mut hashable_parts = self.hashable_parts();
+
+        let (hashed_proposer, _) = hashable_parts
+            .pop()
+            .expect("should contain hashed and serialized proposer");
+        let (hashed_transfer_hashes, _) = hashable_parts
+            .pop()
+            .expect("should contain hashed and serialized transfers");
+        let (hashed_deploy_hashes, _) = hashable_parts
+            .pop()
+            .expect("should contain hashed and serialized deploys");
+        // Make sure we popped everything.
+        assert!(hashable_parts.is_empty());
+
+        hash::hash_slice_rfold(&[
+            hashed_deploy_hashes,
+            hashed_transfer_hashes,
+            hashed_proposer,
+        ])
+    }
+
+    /// Returns a `Vec` of serialized fields along with their hashes.
+    pub(crate) fn hashable_parts(&self) -> Vec<(Digest, Vec<u8>)> {
+        // Pattern match here leverages compiler to ensure every field is accounted for
+        let BlockBody {
+            deploy_hashes,
+            transfer_hashes,
+            proposer,
+        } = self;
+
+        let hashed_deploy_hashes =
+            hash::hash_vec_merkle_tree(deploy_hashes.iter().cloned().map(Into::into).collect());
+        let hashed_transfer_hashes =
+            hash::hash_vec_merkle_tree(transfer_hashes.iter().cloned().map(Into::into).collect());
+        let hashed_proposer =
+            hash::hash(&proposer.to_bytes().expect("Could not serialize proposer"));
+
+        let serialized_deploy_hashes = deploy_hashes
+            .to_bytes()
+            .unwrap_or_else(|error| panic!("should serialize deploy hashes: {}", error));
+        let serialized_transfer_hashes = transfer_hashes
+            .to_bytes()
+            .unwrap_or_else(|error| panic!("should serialize transfer hashes: {}", error));
+        let serialized_proposer = proposer
+            .to_bytes()
+            .unwrap_or_else(|error| panic!("should serialize proposer: {}", error));
+
+        vec![
+            (hashed_deploy_hashes, serialized_deploy_hashes),
+            (hashed_transfer_hashes, serialized_transfer_hashes),
+            (hashed_proposer, serialized_proposer),
+        ]
```

Comment:
Would it be safer to make this a `struct HashableBlockParts` instead of a vec, with named fields `hashed_deploy_hashes` etc., or a wrapper for the `Vec`?
It looks a bit dangerous if we just explicitly access entries `0`, `1` and `2` in different parts of the code (also `storage.rs`).

Reply:
Fixed in 6f40e0e2aec17b2c2aef8885519a7d21df6cc790